controllers/machinepool_controller.go

```diff
@@ -57,13 +76,181 @@ func (r *MachinePoolReconciler) SetupWithManager(mgr ctrl.Manager, options contr
 	r.controller = c
 	r.recorder = mgr.GetEventRecorderFor("machinepool-controller")
 	r.config = mgr.GetConfig()
-
 	r.scheme = mgr.GetScheme()
 	return nil
 }
 
-func (r *MachinePoolReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
-	// TODO(juan-lee): Add machine pool implementation.
+func (r *MachinePoolReconciler) Reconcile(req ctrl.Request) (_ ctrl.Result, reterr error) {
 	// TODO(vincepri): Add support for pause annotation and honoring Cluster.Spec.Paused.
+	ctx := context.Background()
+	_ = r.Log.WithValues("machinepool", req.NamespacedName)
+
+	mp := &clusterv1.MachinePool{}
+	if err := r.Client.Get(ctx, req.NamespacedName, mp); err != nil {
+		if apierrors.IsNotFound(err) {
+			// Object not found, return. Created objects are automatically garbage collected.
+			// For additional cleanup logic use finalizers.
+			return ctrl.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return ctrl.Result{}, err
+	}
+
+	// Initialize the patch helper.
+	patchHelper, err := patch.NewHelper(mp, r.Client)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	defer func() {
+		// Always reconcile the Status.Phase field.
+		r.reconcilePhase(mp)
+
+		// Always attempt to Patch the MachinePool object and status after each reconciliation.
+		if err := patchHelper.Patch(ctx, mp); err != nil {
+			if reterr == nil {
+				reterr = err
+			}
```

Comment:
We can use this pattern to aggregate errors:
`reterr = kerrors.NewAggregate([]error{reterr, err})`

Reply:
without the need for the `if reterr == nil`